Add PHP Core Round #9 (#74)
ThePHPF/thephp.foundation: source/_posts/2023-01-30-php-core-roundup-9.md

*	**RFC Implemented: [Randomizer Additions](https://wiki.php.net/rfc/randomizer_additions)**

	RFC by Joshua Rüsweg and Tim Düsterhus, proposed to add new “building block” methods to `\Random\Randomizer` (added in PHP 8.2) that implement commonly useful operations that are either verbose or very hard to implement in userland.

	The RFC proposed to add two new methods: `getBytesFromString`, to generate a random string containing specific characters, and `getFloat`/`nextFloat`, to generate a random floating point value.

*	**RFC Under Discussion: [Pass Scope to Magic Accessors](https://wiki.php.net/rfc/pass_scope_to_magic_accessors) 💜**

	RFC by Nicolas Grekas and Ilija Tovilo, proposing to pass the calling scope to magic accessors to make it trivial to get it. While it is currently possible to retrieve the calling scope by retrieving the backtrace, the RFC proposes to pass the calling scope as the last argument to the magic methods, making it easier compared to fiddling with the debug backtrace.

*	**RFC Under Discussion: [Path to Saner Increment/Decrement operators](https://wiki.php.net/rfc/saner-inc-dec-operators) 💜**

> 	PHP's increment and decrement operators can have some surprising behaviours when used with types other than int and float. Various previous attempts 1) 2) 3) have been made to improve the behaviour of these operators, but none have been implemented. The goal of this RFC by George Peter Banyard is to normalize the behaviour of `$v++` and `$v--` to be the same as `$v += 1` and `$v -= 1`, respectively.

Reviewer: The "various attempts" with the numbers looks odd.

Author: Thank you. You are right it indeed it was missing the links to the past RFCs. Fixed now.

---
title: 'PHP Core Roundup #9'
layout: post
tags:
    - roundup
author:
  name: Ayesh Karunaratne
  url: https://aye.sh
published_at: 30 January 2023
---

The first for the year 2023, and the ninth in the series, this episode of PHP Core Roundup is full of exciting new developments in PHP and about PHP 8.2, the latest PHP version released last month! January was a busy month with many RFC proposals, discussions, votes, and some that are already included for the upcoming PHP 8.3 version that is scheduled for November 2023.

> The PHP Foundation currently supports six part-time PHP contributors who work on maintenance and new features for PHP. Maintenance is not limited to fixing bugs, but also includes work to reduce technical debt, making life easier for everyone working on PHP. The contributors funded by the PHP Foundation collaborate with other contributors on code, documentation, and discussions.

<br>
We publish the posts on our website, and you can subscribe to a newsletter; You don’t necessarily have to be a PHP Foundation backer to follow _PHP Core Roundup_.

{% include "newsletter.html" %}

## PHP 8.2 Released

> PHP 8.2 is here! With type-system improvements, `readonly `classes, sensitive parameter redaction support, a new "`random`" extension, and several new features and improvements, PHP 8.2.0 was released on December 08th, followed by a bug-fix release 8.2.1 on January 04th.

Reviewer 2: ```suggestion
PHP 8.2 is here! With type-system improvements, `readonly `classes, sensitive parameter redaction support, a new "`random`" extension, and several new features and improvements, PHP 8.2.0 was released on December 08th, followed by a security-fix release 8.2.1 on January 05th.
```

8.2.1  is a [security release](https://github.com/php/web-php/blob/d9a32bf038306184236cd2336620f1143a819f18/include/version.inc#L25)

Over **110 people**, along with our six PHP Foundation members, helped shape PHP 8.2. **Thank you for your amazing efforts 🙏🏼💜**.

[PHP 8.2 Release Announcement](https://www.php.net/releases/8.2/en.php) on [php.net](https://php.net) contains a summary of what’s new and changed in PHP 8.2. Detailed lists and guides are also available on [PHP.Watch: PHP 8.2](https://php.watch/versions/8.2) and [Stitcher.io](https://stitcher.io/blog/new-in-php-82).

## PHP 8.3 RFCs and Discussions

Discussions and proposals for new features and changes for the upcoming PHP version 8.3 were being made even before PHP 8.2.0 was released. January was a particularly eventful month with several proposals and discussions, including a few that went to a vote, and a few that are now implemented in the PHP master branch.

> Most of the proposed improvements are in improving existing PHP functionality, especially the new readonly properties/class and the random extension. There were also discussions and proposals about adding a new syntax similar to Swift to declare granular get/set access to properties, bring back the literal-string RFC, passing the calling scope to class magic methods, and more.

Reviewer 3: Where the did literal string RFC appear? I haven't seen anything like this in the past month.

Author: It's [LiteralString](https://wiki.php.net/rfc/literal_string). I couldn't post a December update, so I thought to include this one from December.

Reviewer 3: Please do not announce drafts, as this hasn't been announced on the mailing list it should be considered a W.I.P.

Author: Understood, makes sense to not include unannounced RFCs. Removed LiteralString.

> Additionally, the [Deprecations for PHP 8.3 RFC](https://wiki.php.net/rfc/deprecations_php_8_3) is also under discussion on features that the community proposes to deprecate in PHP 8.3 and remove in PHP 9.0

Reviewer 3: I wouldn't call this RFC under discussion, it's mainly a W.I.P. draft until this gets officially announced probably around May.

Reviewer: This is marked resolved, but it appears the phrasing wasn't updated.

### RFC Updates

Following are the RFCs and major pull-requests discussed, voted, and implemented since our last update.

*	**RFC In Voting: [Readonly amendments](https://wiki.php.net/rfc/readonly_amendments) 💜**

	RFC by Nicolas Grekas and Máté Kocsis, attempts to address some of the shortcomings of PHP 8.1 readonly properties and 8.2 readonly classes.

	This RFC proposes allowing `readonly` classes to be extended by non-readonly classes (currently not allowed, and causes a fatal error), and to allow reinitializing readonly properties during cloning (within the `__clone()` magic method).

*	**RFC Accepted: [More Appropriate Date/Time Exceptions](https://wiki.php.net/rfc/datetime-exceptions) 💜**

	RFC by Derick Rethans, proposed to introduce Date/Time extension-specific exceptions and errors. This detailed RFC suggests more specificity in the exceptions with exception classes such as `DateInvalidTimeZoneException`, and `DateMalformedPeriodStringException` as well as promoting some of the current PHP warnings to Error exceptions.

>     The vote was accepted, and is pending implementation.

Reviewer 4: It's now further then that, there is a PR that's nearly ready to be merged: https://github.com/php/php-src/pull/10366

Author: Thank you, I will update it.

Author: (and make sure to update the text if the post goes live after the PR is merged).

*	**RFC Under Discussion: [Pass Scope to Magic Accessors](https://wiki.php.net/rfc/pass_scope_to_magic_accessors) 💜**

	RFC by Nicolas Grekas and Ilija Tovilo, proposing to pass the calling scope to magic accessors to make it trivial to get it. While it is currently possible to retrieve the calling scope by retrieving the backtrace, the RFC proposes to pass the calling scope as the last argument to the magic methods, making it easier compared to fiddling with the debug backtrace.

*	**RFC Under Discussion: [Saner array_(sum|product)()](https://wiki.php.net/rfc/saner-array-sum-product) 💜**

	RFC by George Peter Banyard suggests improvements to the existing array_sum and array_product functions by emitting a warning when the arrays being summed/multiplied contain invalid types. This can introduce backwards-incompatible effects (apart from the warning) on code that relies on the current behavior of the functions.

*	**RFC Under Discussion: [Add file_descriptor() function](https://wiki.php.net/rfc/file-descriptor-function) 💜**

> 	RFC by George Peter Banyard proposes a new `file_descriptor()` function that returns the integer file descriptor of a given stream, if the underlying file system supports it.

Reviewer 3: Out of all the ones I've proposed, you skipped the Increment/Decrement one which is probably the bigger/more controversial one. .-.

Also SameSite cookie as a parameter, but that one is small.

Author: Ah good catch, thank you. I will add them.

 
### Notable Mailing List Discussions

* [base64url format](https://externals.io/message/119243)
* [PHP build for the wasm32-wasi target](https://externals.io/message/119198)
* [A set of 18 functions/changes to improve PHP core](https://externals.io/message/119238)
* [PHP support for matrix operations - BLAS, LAPACK](https://externals.io/message/119086)
* [Unicode Text Processing](https://externals.io/message/119149)
* [Deprecate ldap_connect with host and port as separate arguments](https://externals.io/message/119425)
* [Introduce the ability to use the first-call-callable syntax on non-static methods, statically](https://externals.io/message/119392)
* [Loading SQLite extensions on PDO](https://externals.io/message/119374)
* [rules for #include directives](https://externals.io/message/119272)

## Merged PRs and Commits

Following are some changes that did not go through an RFC process because they are either planned, bug fixes, or progressive enhancements. There are automated unit and integration tests for each of these changes,  the PHP core developers review all pull requests.

> 

Reviewer 3: It may make sense to highlight outside of the commit list the improvement in performance of MBString by Alex Dowad.

Author: You are right, they are noteworthy indeed. I mentioned them above the rest of the commit log. I included Alex's work on mbstring, along with the Random extension and const/include RFC mentions.

 - Add a new `imap_is_open()` function to check that a connection object is still valid in [52a891aeaa](https://github.com/php/php-src/commit/52a891aeaa) by George Peter Banyard 💜
 - Add `Randomizer::nextFloat()` and `Randomizer::getFloat()` in [GH-9679](https://github.com/php/php-src/pull/9679) by Tim Düsterhus
 - Implement `mb_substr_count` using fast text conversion filters in [b9cd1cdb4f](https://github.com/php/php-src/commit/b9cd1cdb4f) by Alex Dowad
 - Replace another root XML element format to the "canonical" one in [60cf9fbee0](https://github.com/php/php-src/commit/60cf9fbee0) by Máté Kocsis 💜
 - Remove the superfluous closing parentheses from class synopsis page includes in [b4df038cee](https://github.com/php/php-src/commit/b4df038cee) by Máté Kocsis 💜
 - Always include the constructor on the class manual pages in [0fc60fab72](https://github.com/php/php-src/commit/0fc60fab72) by Máté Kocsis 💜
 - Backport methodsynopsis role attributes changes from master in [6aa5e58414](https://github.com/php/php-src/commit/6aa5e58414) by Máté Kocsis 💜

>  - Fix [GH-10112](https://github.com/php/php-src/issues/10112): LDAP\Connection::__construct() refers to ldap_create() in [GH-10115](https://github.com/php/php-src/pull/10115) by Christoph M. Becker

Reviewer 2: ```suggestion
 - Fix [GH-10112](https://github.com/php/php-src/issues/10112): `LDAP\Connection::__construct()` refers to `ldap_create()` in [GH-10115](https://github.com/php/php-src/pull/10115) by Christoph M. Becker
```

 - Only include the default constructor for non-abstract class synopses in [d832125b8e](https://github.com/php/php-src/commit/d832125b8e) by Máté Kocsis 💜
 - Limit stack size in [GH-9104](https://github.com/php/php-src/pull/9104) by Arnaud Le Blanc 💜
 - Optimize JMP[N]Z_EX to BOOL instead of QM_ASSIGN ([#10108](https://bugs.php.net/bug.php?id=10108)) in [6d9d2eb355](https://github.com/php/php-src/commit/6d9d2eb355) by Ilija Tovilo 💜
 - Remove unnecessary check of p in `phpdbg_trim` ([#10122](https://bugs.php.net/bug.php?id=10122)) in [e288438373](https://github.com/php/php-src/commit/e288438373) by Niels
 - Speed boost for `mb_stripos` (when not using UTF-8) in [744ca16e73](https://github.com/php/php-src/commit/744ca16e73) by Alex Dowad

>  - mb_str{i,}pos does not match illegal byte sequences against occurrences of mb_substitute_char in [GH-9613](https://github.com/php/php-src/pull/9613) by Alex Dowad

Reviewer 2: ```suggestion
 - `mb_str{i,}pos` does not match illegal byte sequences against occurrences of `mb_substitute_char` in [GH-9613](https://github.com/php/php-src/pull/9613) by Alex Dowad
```

>  - Fix [#81740](https://bugs.php.net/bug.php?id=81740): PDO::quote() may return unquoted string in [921b6813da](https://github.com/php/php-src/commit/921b6813da) by Christoph M. Becker

Reviewer 2: ```suggestion
 - Fix [#81740](https://bugs.php.net/bug.php?id=81740): `PDO::quote()` may return unquoted string in [921b6813da](https://github.com/php/php-src/commit/921b6813da) by Christoph M. Becker
```

 - Make build work with newer OpenSSL in [5f90134bb6](https://github.com/php/php-src/commit/5f90134bb6) by Stanislav Malyshev
 - Improve fix for bug [#81740](https://bugs.php.net/bug.php?id=81740) in [a6a80eefe0](https://github.com/php/php-src/commit/a6a80eefe0) by Stanislav Malyshev
 - Fix memory leak in [683285165e](https://github.com/php/php-src/commit/683285165e) by Dmitry Stogov
 - `mbstring`: Do not stop when mbstring test failed in [e0e587cdb8](https://github.com/php/php-src/commit/e0e587cdb8) by Yuya Hamada
 - Skip newly added test on 32bit platforms in [cf5dac07d2](https://github.com/php/php-src/commit/cf5dac07d2) by Christoph M. Becker
 - Force extension loading for new test in [da5cbca23e](https://github.com/php/php-src/commit/da5cbca23e) by Christoph M. Becker
 - Skip tests under MSAN in [cf77762970](https://github.com/php/php-src/commit/cf77762970) by Arnaud Le Blanc 💜
 - Fix [GH-9891](https://github.com/php/php-src/issues/9891): DateTime modify with unixtimestamp (@) must work like setTimestamp in [d19a70c9a0](https://github.com/php/php-src/commit/d19a70c9a0) by Derick Rethans 💜

>  - Fix [#10133](https://bugs.php.net/bug.php?id=10133) set variables_order en ensure $ENV is set in [bfa56cf62b](https://github.com/php/php-src/commit/bfa56cf62b) by Remi Collet

Reviewer 2: ```suggestion
 - Fix [#10133](https://bugs.php.net/bug.php?id=10133) set variables_order en ensure `$ENV` is set in [bfa56cf62b](https://github.com/php/php-src/commit/bfa56cf62b) by Remi Collet
```

 - Remove `_zend_ast_decl->lex_pos` ([#10046](https://bugs.php.net/bug.php?id=10046)) in [23fe58c3a8](https://github.com/php/php-src/commit/23fe58c3a8) by Niels
 - cli server addressing few todos in [GH-10124](https://github.com/php/php-src/pull/10124) by David Carlier
 - sockets adding `TCP_QUICKACK` constant in [GH-10145](https://github.com/php/php-src/pull/10145) by David Carlier
 - Add a regression test for `auto_globals_jit=0` with preloading on in [bbad29b9c1](https://github.com/php/php-src/commit/bbad29b9c1) by Niels Dossche
 - Fix undefined behaviour in `phpdbg_load_module_or_extension` in [GH-10157](https://github.com/php/php-src/pull/10157) by Niels Dossche
 - Update test for changed behaviour of GMP constructor in [a24659e70c](https://github.com/php/php-src/commit/a24659e70c) by Niels Dossche
 - Fix [GH-10072](https://github.com/php/php-src/issues/10072): PHP crashes when execute_ex is overridden and a __call trampoline is used from internal code in [233ffccc35](https://github.com/php/php-src/commit/233ffccc35) by Derick Rethans 💜

>  - Make sure to disable JIT when overriding execute_ex in [b489e0f2b8](https://github.com/php/php-src/commit/b489e0f2b8) by Derick Rethans 💜

Reviewer 4: That's in a test case.

 - Fix [GH-9981](https://github.com/php/php-src/issues/9981): FPM does not reset fastcgi.error_header in [a3891d9d1a](https://github.com/php/php-src/commit/a3891d9d1a) by Jakub Zelenka 💜
 - Fix memory leak because of incorrect optimization in [0464524292](https://github.com/php/php-src/commit/0464524292) by Dmitry Stogov
 - Fix bug [#68591](https://bugs.php.net/bug.php?id=68591): Configuration test does not perform UID lookups in [GH-10165](https://github.com/php/php-src/pull/10165) by Jakub Zelenka 💜
 - ext/opcache/jit: handle `zend_jit_find_trace()` failures in [b26b758952](https://github.com/php/php-src/commit/b26b758952) by Max Kellermann
 - Use proper int|float union type instead of numeric ([#10162](https://bugs.php.net/bug.php?id=10162)) in [4cee2c0127](https://github.com/php/php-src/commit/4cee2c0127) by George Peter Banyard 💜
 - Added missed return in [ca5f668f7c](https://github.com/php/php-src/commit/ca5f668f7c) by Dmitry Stogov
 - ext/opcache/jit/zend_jit_trace: add missing lock for EXIT_INVALIDATE in [e217138b40](https://github.com/php/php-src/commit/e217138b40) by Max Kellermann

>  - Allow 'h' and 'k' flags to be combined for mb_convert_kana in [GH-10174](https://github.com/php/php-src/pull/10174) by Alex Dowad

Reviewer 2: ```suggestion
 - Allow `h` and `k` flags to be combined for `mb_convert_kana` in [GH-10174](https://github.com/php/php-src/pull/10174) by Alex Dowad
```

 - Register parameter attributes via stub in `ext/zend_test` ([#10183](https://bugs.php.net/bug.php?id=10183)) in [3e48e52d93](https://github.com/php/php-src/commit/3e48e52d93) by Tim Düsterhus
 - Fix null pointer dereference of param in [3a44c78f14](https://github.com/php/php-src/commit/3a44c78f14) by Niels Dossche

>  - Improve mb_detect_encoding's recognition of Turkish text in [f40c3fca88](https://github.com/php/php-src/commit/f40c3fca88) by Alex Dowad

Reviewer 2: ```suggestion
 - Improve `mb_detect_encoding`'s recognition of Turkish text in [f40c3fca88](https://github.com/php/php-src/commit/f40c3fca88) by Alex Dowad
```

>  - Fix memory leak in posix_ttyname() in [GH-10190](https://github.com/php/php-src/pull/10190) by George Peter Banyard 💜

Reviewer 2: ```suggestion
 - Fix memory leak in `posix_ttyname()` in [GH-10190](https://github.com/php/php-src/pull/10190) by George Peter Banyard 💜
```

 - Do not display non-existent constructors in [38e138798d](https://github.com/php/php-src/commit/38e138798d) by Máté Kocsis 💜
 - Do not list private constructors as inherited in [7b08fe9f2d](https://github.com/php/php-src/commit/7b08fe9f2d) by Máté Kocsis 💜
 - Fix [GH-10200](https://github.com/php/php-src/issues/10200): zif_get_object_vars: Assertion `!(((__ht)->u.flags & (1<<2)) != 0)' failed in [GH-10209](https://github.com/php/php-src/pull/10209) by Niels Dossche
 - Fix [GH-10200](https://github.com/php/php-src/issues/10200): zif_get_object_vars: Assertion `!(((__ht)->u.flags & (1<<2)) != 0)' failed in [GH-10209](https://github.com/php/php-src/pull/10209) by Niels Dossche
 - ext/opcache/zend_shared_alloc: add assertions on "locked" flag in [e1a25ff2ed](https://github.com/php/php-src/commit/e1a25ff2ed) by Max Kellermann
 - ext/opcache/zend_shared_alloc: change "locked" check to assertion in [10d43c40dd](https://github.com/php/php-src/commit/10d43c40dd) by Max Kellermann
 - Fix [GH-10202](https://github.com/php/php-src/issues/10202): posix_getgr(gid|nam)_basic.phpt fail in [d5f0362e59](https://github.com/php/php-src/commit/d5f0362e59) by Niels Dossche

>  - Implement mb_output_handler using fast text conversion filters in [a9a672048b](https://github.com/php/php-src/commit/a9a672048b) by Alex Dowad

Reviewer 2: ```suggestion
 - Implement `mb_output_handler` using fast text conversion filters in [a9a672048b](https://github.com/php/php-src/commit/a9a672048b) by Alex Dowad
```

>  - Implement mb_str_split using fast text conversion filters in [88c99afdac](https://github.com/php/php-src/commit/88c99afdac) by Alex Dowad

Reviewer 2: ```suggestion
 - Implement `mb_str_split` using fast text conversion filters in [88c99afdac](https://github.com/php/php-src/commit/88c99afdac) by Alex Dowad
```

 - Mark default interned strings as valid UTF-8 where appropriate in [e2654a532a](https://github.com/php/php-src/commit/e2654a532a) by Alex Dowad

>  - Implement php_mb_zend_encoding_converter using fast text conversion filters in [953864661a](https://github.com/php/php-src/commit/953864661a) by Alex Dowad

Reviewer 2: ```suggestion
 - Implement `php_mb_zend_encoding_converter` using fast text conversion filters in [953864661a](https://github.com/php/php-src/commit/953864661a) by Alex Dowad
```

>  - Implement mb_detect_encoding using fast text conversion filters in [0e7160b836](https://github.com/php/php-src/commit/0e7160b836) by Alex Dowad

Reviewer 2: ```suggestion
 - Implement `mb_detect_encoding` using fast text conversion filters in [0e7160b836](https://github.com/php/php-src/commit/0e7160b836) by Alex Dowad
```

>  - Use smart_str in mb_http_input rather than mbfl_memory_device in [3b5072f6f6](https://github.com/php/php-src/commit/3b5072f6f6) by Alex Dowad

Reviewer 2: ```suggestion
 - Use `smart_str` in `mb_http_input` rather than `mbfl_memory_device` in [3b5072f6f6](https://github.com/php/php-src/commit/3b5072f6f6) by Alex Dowad
```

 - Move test for [GH-10200](https://github.com/php/php-src/issues/10200) to the simplexml extension test directory in [GH-10252](https://github.com/php/php-src/pull/10252) by Niels Dossche
 - random: Fix check before closing `random_fd` ([#10247](https://bugs.php.net/bug.php?id=10247)) in [32f503e4e3](https://github.com/php/php-src/commit/32f503e4e3) by Tim Düsterhus
 - Remove unnecessary NULL-checks on ctx ([#10256](https://bugs.php.net/bug.php?id=10256)) in [58d741c042](https://github.com/php/php-src/commit/58d741c042) by Niels
 - Remove 'fast path' using mblen_table from mb_get_strlen (it's actually a slow path) in [cca4ca6d3d](https://github.com/php/php-src/commit/cca4ca6d3d) by Alex Dowad
 - Add unit tests for mb_str_split/mb_substr on MacJapanese encoding in [d8b5b9fa55](https://github.com/php/php-src/commit/d8b5b9fa55) by Alex Dowad
 - Optimize branch structure of UTF-8 decoder routine in [092ad3e462](https://github.com/php/php-src/commit/092ad3e462) by Alex Dowad
 - Fix [GH-9710](https://github.com/php/php-src/issues/9710): phpdbg memory leaks by option "-h" in [GH-10237](https://github.com/php/php-src/pull/10237) by Niels Dossche

>  - Fix [GH-10251](https://github.com/php/php-src/issues/10251): Assertion `(flag & (1<<3)) == 0' failed in [GH-10254](https://github.com/php/php-src/pull/10254) by Niels Dossche

Reviewer 2: ```suggestion
 - Fix [GH-10251](https://github.com/php/php-src/issues/10251): Assertion `(flag & (1<<3)) == 0` failed in [GH-10254](https://github.com/php/php-src/pull/10254) by Niels Dossche
```

 - Zend/zend_smart_str: include cleanup in [738fb5ca54](https://github.com/php/php-src/commit/738fb5ca54) by Max Kellermann
 - main: add missing includes in [16203b53e1](https://github.com/php/php-src/commit/16203b53e1) by Max Kellermann
 - ext/{standard,json,random,...}: add missing includes in [308fd311ea](https://github.com/php/php-src/commit/308fd311ea) by Max Kellermann
 - Zend/zend_fibers: include cleanup in [aa1cd02a43](https://github.com/php/php-src/commit/aa1cd02a43) by Max Kellermann
 - sapi/*: add missing includes in [6b34de8eba](https://github.com/php/php-src/commit/6b34de8eba) by Max Kellermann
 - Zend/zend_{operators,variables}: include cleanup in [694ec1deea](https://github.com/php/php-src/commit/694ec1deea) by Max Kellermann
 - Zend/zend_object_handlers: include cleanup in [b4ba16fe18](https://github.com/php/php-src/commit/b4ba16fe18) by Max Kellermann
 - Zend/zend_ptr_stack: include cleanup in [f377e15751](https://github.com/php/php-src/commit/f377e15751) by Max Kellermann
 - Zend/zend_multibyte: include cleanup in [588a07f737](https://github.com/php/php-src/commit/588a07f737) by Max Kellermann

>  - Zend/zend_execute: include cleanup in [ecc880f491](https://github.com/php/php-src/commit/ecc880f491) by Max Kellermann

Reviewer 3: All of the include cleanup commits got reverted.

 - Fix incorrect trace type inference when function SSA is not available in [298aa74b16](https://github.com/php/php-src/commit/298aa74b16) by Dmitry Stogov
 - Revert "`ext/opcache`: use C11 atomics for "restart_in" ([#10276](https://bugs.php.net/bug.php?id=10276))" ([#10339](https://bugs.php.net/bug.php?id=10339)) in [bff7a56d00](https://github.com/php/php-src/commit/bff7a56d00) by Max Kellermann
 - Sync with timelib 2021.19 in [ce877da23b](https://github.com/php/php-src/commit/ce877da23b), [4d8f981818](https://github.com/php/php-src/commit/4d8f981818), and [0df92d218e](https://github.com/php/php-src/commit/0df92d218e) by Derick Rethans 💜
 - Fix [GH-8086](https://github.com/php/php-src/issues/8086): Introduce mail.mixed_lf_and_crlf INI in [cc931af35d](https://github.com/php/php-src/commit/cc931af35d) by Jakub Zelenka 💜
 - Delay debug JIT op_array dump until actual JIT-ing in [6676f5d3cb](https://github.com/php/php-src/commit/6676f5d3cb) by Dmitry Stogov
 - Tweak SSE2-accelerated strtoupper() and strtolower() for speed in [0b7986f976](https://github.com/php/php-src/commit/0b7986f976) by Alex Dowad

>  - Fix [GH-10248](https://github.com/php/php-src/issues/10248): Assertion `!(zval_get_type(&(*(property))) == 10)' failed in [0801c567dc](https://github.com/php/php-src/commit/0801c567dc) by Niels Dossche

Reviewer 2: ```suggestion
 - Fix [GH-10248](https://github.com/php/php-src/issues/10248): Assertion `!(zval_get_type(&(*(property))) == 10)` failed in [0801c567dc](https://github.com/php/php-src/commit/0801c567dc) by Niels Dossche
```
